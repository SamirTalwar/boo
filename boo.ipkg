package boo

sourcedir = "proofs"

modules =
    Builtins
  , EvaluationReduction
  , Expression
  , Identifier
  , Native
  , Primitive
  , Test
