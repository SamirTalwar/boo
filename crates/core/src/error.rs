use crate::span::Span;

pub type Result<T> = std::result::Result<T, Error>;

#[derive(Debug, Clone, PartialEq, thiserror::Error, miette::Diagnostic)]
pub enum Error {
    #[error("Unexpected token: {token}")]
    #[diagnostic(code(boo::lex::unexpected_token))]
    UnexpectedToken {
        #[label("unexpected token")]
        span: Span,
        token: String,
    },

    #[error("Parse error: expected one of {expected_tokens:?}")]
    #[diagnostic(code(boo::parse::error))]
    ParseError {
        #[label("{}", expected_one_of(expected_tokens))]
        span: Span,
        expected_tokens: Vec<&'static str>,
    },

    #[error("Unknown variable: {name:?}")]
    #[diagnostic(code(boo::interpret::unknown_variable))]
    UnknownVariable {
        #[label("unknown variable")]
        span: Span,
        name: String,
    },

    #[error("Could not apply the function")]
    #[diagnostic(code(boo::interpret::unknown_variable))]
    InvalidFunctionApplication {
        #[label("invalid function")]
        span: Span,
    },
}

fn expected_one_of(strings: &[&str]) -> String {
    match strings {
        [] => "<nothing>".to_string(),
        [a] => format!("expected {}", a),
        [a, b] => format!("expected {} or {}", a, b),
        _ => format!(
            "expected one of {}, or {}",
            strings[0..strings.len() - 1].join(", "),
            strings[strings.len() - 1]
        ),
    }
}
