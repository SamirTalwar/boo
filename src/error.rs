#[derive(Debug, PartialEq)]
pub enum BooError {
    ParseError(peg::error::ParseError<peg::str::LineCol>),
}

impl std::fmt::Display for BooError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::ParseError(error) => write!(f, "Parse error: {}", error),
        }
    }
}
