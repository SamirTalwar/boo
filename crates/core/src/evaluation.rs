use crate::ast::Function;
use crate::error::Result;
use crate::expr::Expr;
use crate::primitive::Primitive;

pub trait Evaluator {
    fn evaluate(&self, expr: Expr) -> Result<Evaluated>;
}

/// An evaluation result. This can be either a primitive value or a closure.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Evaluated {
    Primitive(Primitive),
    Function(Function<Expr>),
}

impl std::fmt::Display for Evaluated {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Evaluated::Primitive(x) => x.fmt(f),
            Evaluated::Function(x) => x.fmt(f),
        }
    }
}
