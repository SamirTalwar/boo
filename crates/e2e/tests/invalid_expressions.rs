use boo::error::{Error, Result};
use boo::evaluation::Evaluator;
use boo::*;
use boo_evaluation_reduction::ReducingEvaluator;
use boo_optimized_evaluator::PoolingEvaluator;

#[test]
fn test_unknown_variable() -> Result<()> {
    expect_error(
        "unknown_variable",
        "123 + xyz",
        Error::UnknownVariable {
            span: Some((6..9).into()),
            name: "xyz".to_string(),
        },
    )
}

#[test]
fn test_does_not_close_over_variables_out_of_scope() -> Result<()> {
    expect_error(
        "does_not_close_over_variables_out_of_scope",
        "let fun = (let one = 1 in fn param -> one + param + external) in let external = 2 in fun 3",
        Error::UnknownVariable {
            span: Some((52..60).into()),
            name: "external".to_string(),
        },
    )
}

fn expect_error(name: &str, program: &str, expected_error: Error) -> Result<()> {
    let ast = parse(program)?.to_core()?;
    insta::with_settings!({ description => program }, {
        insta::assert_debug_snapshot!(name.to_string() + "__parse", ast);
    });

    let type_check_result = boo_types_hindley_milner::type_of(&ast);
    assert_eq!(type_check_result, Err(expected_error.clone()));

    {
        let mut reducing_evaluator = ReducingEvaluator::new();
        builtins::prepare(&mut reducing_evaluator)?;
        let actual_result = reducing_evaluator.evaluate(ast.clone());
        assert_eq!(actual_result, Err(expected_error.clone()));
    }

    {
        let mut optimized_evaluator = PoolingEvaluator::new_recursive();
        builtins::prepare(&mut optimized_evaluator)?;
        let actual_result = optimized_evaluator.evaluate(ast);
        assert_eq!(actual_result, Err(expected_error));
    }

    Ok(())
}
