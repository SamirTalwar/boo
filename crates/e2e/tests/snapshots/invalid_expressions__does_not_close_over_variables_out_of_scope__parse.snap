---
source: crates/e2e/tests/invalid_expressions.rs
description: let fun = (let one = 1 in fn param -> one + param + external) in let external = 2 in fun 3
expression: ast
---
Expr {
    span: Some(
        Span {
            start: 0,
            end: 90,
        },
    ),
    expression: Assign(
        Assign {
            name: Name(
                "fun",
            ),
            value: Expr {
                span: Some(
                    Span {
                        start: 11,
                        end: 60,
                    },
                ),
                expression: Assign(
                    Assign {
                        name: Name(
                            "one",
                        ),
                        value: Expr {
                            span: Some(
                                Span {
                                    start: 21,
                                    end: 22,
                                },
                            ),
                            expression: Primitive(
                                Integer(
                                    Small(
                                        1,
                                    ),
                                ),
                            ),
                        },
                        inner: Expr {
                            span: Some(
                                Span {
                                    start: 26,
                                    end: 60,
                                },
                            ),
                            expression: Function(
                                Function {
                                    parameter: Name(
                                        "param",
                                    ),
                                    body: Expr {
                                        span: Some(
                                            Span {
                                                start: 38,
                                                end: 60,
                                            },
                                        ),
                                        expression: Apply(
                                            Apply {
                                                function: Expr {
                                                    span: Some(
                                                        Span {
                                                            start: 38,
                                                            end: 60,
                                                        },
                                                    ),
                                                    expression: Apply(
                                                        Apply {
                                                            function: Expr {
                                                                span: Some(
                                                                    Span {
                                                                        start: 38,
                                                                        end: 60,
                                                                    },
                                                                ),
                                                                expression: Identifier(
                                                                    Operator(
                                                                        "+",
                                                                    ),
                                                                ),
                                                            },
                                                            argument: Expr {
                                                                span: Some(
                                                                    Span {
                                                                        start: 38,
                                                                        end: 49,
                                                                    },
                                                                ),
                                                                expression: Apply(
                                                                    Apply {
                                                                        function: Expr {
                                                                            span: Some(
                                                                                Span {
                                                                                    start: 38,
                                                                                    end: 49,
                                                                                },
                                                                            ),
                                                                            expression: Apply(
                                                                                Apply {
                                                                                    function: Expr {
                                                                                        span: Some(
                                                                                            Span {
                                                                                                start: 38,
                                                                                                end: 49,
                                                                                            },
                                                                                        ),
                                                                                        expression: Identifier(
                                                                                            Operator(
                                                                                                "+",
                                                                                            ),
                                                                                        ),
                                                                                    },
                                                                                    argument: Expr {
                                                                                        span: Some(
                                                                                            Span {
                                                                                                start: 38,
                                                                                                end: 41,
                                                                                            },
                                                                                        ),
                                                                                        expression: Identifier(
                                                                                            Name(
                                                                                                "one",
                                                                                            ),
                                                                                        ),
                                                                                    },
                                                                                },
                                                                            ),
                                                                        },
                                                                        argument: Expr {
                                                                            span: Some(
                                                                                Span {
                                                                                    start: 44,
                                                                                    end: 49,
                                                                                },
                                                                            ),
                                                                            expression: Identifier(
                                                                                Name(
                                                                                    "param",
                                                                                ),
                                                                            ),
                                                                        },
                                                                    },
                                                                ),
                                                            },
                                                        },
                                                    ),
                                                },
                                                argument: Expr {
                                                    span: Some(
                                                        Span {
                                                            start: 52,
                                                            end: 60,
                                                        },
                                                    ),
                                                    expression: Identifier(
                                                        Name(
                                                            "external",
                                                        ),
                                                    ),
                                                },
                                            },
                                        ),
                                    },
                                },
                            ),
                        },
                    },
                ),
            },
            inner: Expr {
                span: Some(
                    Span {
                        start: 65,
                        end: 90,
                    },
                ),
                expression: Assign(
                    Assign {
                        name: Name(
                            "external",
                        ),
                        value: Expr {
                            span: Some(
                                Span {
                                    start: 80,
                                    end: 81,
                                },
                            ),
                            expression: Primitive(
                                Integer(
                                    Small(
                                        2,
                                    ),
                                ),
                            ),
                        },
                        inner: Expr {
                            span: Some(
                                Span {
                                    start: 85,
                                    end: 90,
                                },
                            ),
                            expression: Apply(
                                Apply {
                                    function: Expr {
                                        span: Some(
                                            Span {
                                                start: 85,
                                                end: 88,
                                            },
                                        ),
                                        expression: Identifier(
                                            Name(
                                                "fun",
                                            ),
                                        ),
                                    },
                                    argument: Expr {
                                        span: Some(
                                            Span {
                                                start: 89,
                                                end: 90,
                                            },
                                        ),
                                        expression: Primitive(
                                            Integer(
                                                Small(
                                                    3,
                                                ),
                                            ),
                                        ),
                                    },
                                },
                            ),
                        },
                    },
                ),
            },
        },
    ),
}
